templates/service-worker.js

```diff
@@ -94,3 +94,123 @@ self.addEventListener('fetch', event => {
     return fetch(req)
   })())
 })
+
+self.addEventListener('push', event => {
+  event.waitUntil((async () => {
+    const data = event.data.json()
+    const { origin } = new URL(data.icon)
+
+    try {
+      const notification = await get(`${origin}/api/v1/notifications/${data.notification_id}`, {
+        'Authorization': `Bearer ${data.access_token}`,
+      }, { timeout: 2000 })
+
+      await showRichNotification(data, notification)
+    } catch (e) {
+      await showSimpleNotification(data)
+    }
+  })())
+})
+
+async function showSimpleNotification(data) {
+  await self.registration.showNotification(data.title, {
+    icon: data.icon,
+    body: data.body
+  })
+}
+
+async function showRichNotification(data, notification) {
+  switch (notification.type) {
+    case 'follow': {
+      await self.registration.showNotification(data.title, {
+        icon: data.icon,
+        body: data.body,
+        data: {
+          url: `${self.location.origin}/accounts/${notification.account.id}`
+        }
+      })
+      break
+    }
+    case 'mention': {
+      await self.registration.showNotification(data.title, {
+        icon: data.icon,
+        body: data.body,
+        data: {
+          url: `${self.location.origin}/statuses/${notification.status.id}`
+        }
+      })
+      break
+    }
+    case 'reblog': {
+      await self.registration.showNotification(data.title, {
+        icon: data.icon,
+        body: data.body,
+        data: {
+          url: `${self.location.origin}/statuses/${notification.status.id}`
+        }
+      })
+      break
+    }
+    case 'favourite': {
+      await self.registration.showNotification(data.title, {
+        icon: data.icon,
+        body: data.body,
+        data: {
+          url: `${self.location.origin}/statuses/${notification.status.id}`
+        }
+      })
+      break
+    }
+  }
+}
+
+self.addEventListener('notificationclick', event => {
+  event.waitUntil((async () => {
+    if (event.notification.data && event.notification.data.url) {
+      await self.clients.openWindow(event.notification.data.url)
+      await event.notification.close()
+    }
+  })())
+})
+
+// Copy-paste from ajax.js
```

Comment:
Maybe it would make sense to process the file with Webpack so that imports work?

Reply:
SW will be handled by Webpack once I upgrade us to the latest version of Sapper, but unfortunately I'm blocked on that for now: https://github.com/nolanlawson/pinafore/issues/416